repeat_expansion_variants/pipeline.py

```diff
@@ -0,0 +1,209 @@
+        # cycle were allowed to run). This is similar to SQL merge.
+        annotation_df = pd.merge(variants_original, annotation_info, on=column_name_in_dataframe, how='left')
+        # Update main dataframe with the new values. This replaces the NaN values in the dataframe with the ones
+        # available in another dataframe we just created, `annotation_df`.
+        variants = variants \
+            .set_index([column_name_in_dataframe]) \
+            .combine_first(annotation_df.set_index([column_name_in_dataframe]))
+
+    # Reset index to default
+    variants.reset_index(inplace=True)
+    # Some records are being annotated to multiple Ensembl genes. For example, HGNC:10560 is being resolved to
+    # ENSG00000285258 and ENSG00000163635. We need to explode dataframe by that column.
+    variants = variants.explode('EnsemblGeneID')
+    # Fetch Ensembl gene name based on Ensembl gene ID
+    annotation_info = biomart.query_biomart(
+        key_column=('ensembl_gene_id', 'EnsemblGeneID'),
+        query_column=('external_gene_name', 'EnsemblGeneName'),
+        identifier_list=sorted({str(i) for i in variants['EnsemblGeneID'] if str(i).startswith('ENSG')}),
+    )
+    variants = pd.merge(variants, annotation_info, on='EnsemblGeneID', how='left')
+    # Check that there are no multiple gene name mappings for any given gene ID
+    assert variants['EnsemblGeneName'].str.len().dropna().max() == 1, 'Multiple gene ID → gene name mappings found!'
+    # Convert the one-item list into a plain column
+    return variants.explode('EnsemblGeneName')
+
+
+def determine_repeat_type(row):
+    """
+    Based on all available information about a variant, determine its type. The resulting type can be:
+        * trinucleotide_repeat_expansion, corresponding to SO:0002165
+        * short_tandem_repeat_expansion, corresponding to SO:0002162
+        * NaN (not able to determine)
+    Also, depending on the information, determine whether the record is complete, i.e., whether it has all necessary
+    fields to be output for the final "consequences" table.
+    """
+    repeat_type = np.nan
+    if row['IsProteinHGVS']:
+        # For protein HGVS notation, assume that repeat is a trinucleotide one, since it affects entire amino acids
+        repeat_type = 'trinucleotide_repeat_expansion'
+    else:
+        # As a priority, use the repeat unit length determined directly from base sequence
+        repeat_unit_length = row['RepeatUnitLength']
+        # If not available, fall back to using and end coordinate difference
+        if pd.isnull(repeat_unit_length):
+            repeat_unit_length = row['CoordinateSpan']
+        # Determine repeat type based on repeat unit length
+        if pd.notnull(repeat_unit_length):
+            if repeat_unit_length % 3 == 0:
+                repeat_type = 'trinucleotide_repeat_expansion'
+            else:
+                repeat_type = 'short_tandem_repeat_expansion'
```

Comment:
So a five nucleotide repeat expansion would also qualify as a **short_tandem_repeat_expansion**?

Reply:
Exactly. Trinucleotide repeat expansions are a subclass of short tandem repeat expansion. So if it's 3 (or a multiple of 3 depending on interpretation — will discuss with OT directors as mentioned above), it's a `trinucleotide_repeat_expansion`; if it's 5, or 1, or any number of nucleotides, that will get the general calss `short_tandem_repeat_expansion`.